pageserver/src/tenant/timeline.rs

```diff
@@ -3671,6 +3713,15 @@ impl Timeline {
                     "keeping {} because it is the latest layer",
                     l.filename().file_name()
                 );
+                // Collect delta key ranges that need image layers to allow garbage
+                // collecting the layers.
+                // It is not so obvious whether we need to propagate information only about
+                // delta layers. Image layers can form "stairs" preventing old image from been deleted.
+                // But image layers are in any case less sparse than delta layers. Also we need some
+                // protection from replacing recent image layers with new one after each GC iteration.
+                if l.is_incremental() && !LayerMap::is_l0(&*l) {
+                    wanted_image_layers.add_range(l.get_key_range());
```

Comment:
This is a very eager GC approach. Is there a case where we'd rather wait? One example I can think of:
1. User updates rows sequentially, but skipping many rows. We get sparse deltas covering ranges (0GB - 100GB), (100GB, 200GB), (200GB, 300GB), etc.
2. These deltas pass the GC horizon. We create wanted image layers.
3. We create 100GB of images for each delta. Oops, we wrote 100GB of data to delete 0.25GB of data. We reimaged the entire database to delete only a few deltas

Reply:
Once again, I agree with you. But please take in account that:
1. Right scenario you have described is also possible. When we create 3 such wide delta layers then we also trigger generation of image layers covering them. So we also write 100GB of data for 3*0.25GB of changes.
2. Garbage collection is expected to be performed quite rarely. With PiTR=7 days there is not so much sense to try to perform GC more than once per day. Certainly if PiTR will be decreased to 1 hours, then situation will be changed.
3. This mechanism with wanted layers propagation was proposed not only to reclaim space by GC, but also as some additional factor which force generation of image layers and so reducing number of page reconstructions. So writing 100GB of data for  0.25GB of changes is not always bad idea (if it i not performed too frequently). It allows to increase speed of `get_page_at_lsn`.